src/openvic-simulation/tech/Technology.cpp

```diff
@@ -0,0 +1,154 @@
+	if (identifier.empty()) {
+		Logger::error("Invalid technology identifier - empty!");
+		return false;
+	}
+
+	if (area == nullptr) {
+		Logger::error("Null area for technology \"", identifier, "\"!");
+		return false;
+	}
+
+	return technologies.add_item({ identifier, *area, year, cost, unciv_military, unit, activated_units, activated_buildings, std::move(values) });
+}
+
+bool TechnologyManager::add_technology_school(std::string_view identifier, ModifierValue&& values) {
+	if (identifier.empty()) {
+		Logger::error("Invalid modifier effect identifier - empty!");
+		return false;
+	}
+
+	return technology_schools.add_item({ identifier, std::move(values) });
+}
+
+bool TechnologyManager::load_technology_file_areas(ast::NodeCPtr root) {
+	return expect_dictionary([this](std::string_view root_key, ast::NodeCPtr root_value) -> bool {
+		if (root_key == "folders") {
+			return expect_dictionary([this](std::string_view folder_key, ast::NodeCPtr folder_value) -> bool {
+				bool ret = add_technology_folder(folder_key);
+				TechnologyFolder const* current_folder = get_technology_folder_by_identifier(folder_key);
+				if (current_folder == nullptr)
+					return false;
+				ret &= expect_list(expect_identifier([this, current_folder](std::string_view area) -> bool {
+					return add_technology_area(area, current_folder);
+				}))(folder_value);
+				return ret;
+			})(root_value);
+			lock_technology_folders();
+			lock_technology_areas();
+		} else if (root_key == "schools") return true; //ignore
+		else return false;
+	})(root);
+}
+
+bool TechnologyManager::load_technology_file_schools(ModifierManager const& modifier_manager, ast::NodeCPtr root) {
+	return expect_dictionary([this, &modifier_manager](std::string_view root_key, ast::NodeCPtr root_value) -> bool {
+		if (root_key == "schools") {
+			return expect_dictionary([this, &modifier_manager](std::string_view school_key, ast::NodeCPtr school_value) -> bool {
+				ModifierValue modifiers;
+				bool ret = modifier_manager.expect_modifier_value(move_variable_callback(modifiers))(school_value);
+				ret &= add_technology_school(school_key, std::move(modifiers));
+				return ret;
+			})(root_value);
+			lock_technology_schools();
+		} else if (root_key == "folders") return true; //ignore
+		else return false;
+	})(root);
+}
+
+bool TechnologyManager::load_technologies_file(ModifierManager const& modifier_manager, UnitManager const& unit_manager, BuildingManager const& building_manager, ast::NodeCPtr root) {
+	size_t expected_technologies = 0;
+	bool ret = expect_dictionary_reserve_length(technologies, [this, &expected_technologies](std::string_view key, ast::NodeCPtr value) -> bool {
+		return expect_length(add_variable_callback(expected_technologies))(value);
+	})(root);
+	technologies.reserve(technologies.size() + expected_technologies);
+
+	ret &= expect_dictionary([this, &modifier_manager, &unit_manager, &building_manager](std::string_view tech_key, ast::NodeCPtr tech_value) -> bool {
```

Comment:
```suggestion
	return expect_dictionary_reserve_length(technologies, [this, &modifier_manager, &unit_manager, &building_manager](std::string_view tech_key, ast::NodeCPtr tech_value) -> bool {
```
The `expected_*` counting pattern is used to count things spread across multiple parent dictionaries, e.g. count cultures from every culture group. In this case, all the techs are in the top level dictionary of the file (`root`), so we can just reserve its length normally in order to reserve enough space for every tech in the file.
(Also remove `return ret;` at the bottom of this function, now that we only have a single result to return.)